### PR TITLE
feat(filter-field): Add multiselect

### PR DESCRIPTION
**Pull request by subarroca**
#### Type of PR
Feature (non-breaking change which adds functionality)

Added multi select panel for filter field

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: apps/components-e2e/src/components/filter-field/filter-field.e2e.ts [hunk 1]
```diff
@@ -17,20 +17,24 @@
...
   setupSecondTestScenario,
+  switchToSecondDatasource,
+  clickMultiSelectOption,
+  input,
+  errorBox,
+  tagOverlay,
   filterFieldRangePanel,
 } from './filter-field.po';
-import { Selector } from 'testcafe';
-import { waitForAngular, resetWindowSizeToDefault } from '../../utils';
```
Comment (tomheller) [inline apps/components-e2e/src/components/filter-field/filter-field.e2e.ts:33]: This imports have primarily resorted. It would be great to disable automatic import sorting. These changesets are really hard to grasp during a review. Maybe you can clean them up?

File: libs/barista-components/filter-field/README.md [hunk 2]
```diff
@@ -101,6 +101,10 @@ Every filter is an array of:
   user. This includes the range operator (`DtFilterFieldRangeOperator`), the
   unit (`string`) of the selected values and the range itself (which can either
   be one `number` value or a `number` tuple).
+- Multi select filters: array of objects that have been selected via an multi
+  selct list with checkboxes (the objects are exactly the ones the consumer has
```
Comment (tomheller) [inline libs/barista-components/filter-field/README.md:105]:
```suggestion
  select list with checkboxes (the objects are exactly the ones the consumer has
```

File: libs/barista-components/filter-field/README.md [hunk 3]
```diff
@@ -184,11 +188,20 @@ node objects:
...
+| `dtMultiSelectDef`    | Creates a node definition object or extends one and applies a multi select definition object based on the parameters.  |
+| `isDtMultiSelectDef`  | Whether the provided definition object is of type `NodeDef` and consists of a multi select definition.                 |
 
 ## Behavior
 
 <ba-ux-snippet name="filter-field-behavior"></ba-ux-snippet>
 
+### Multi selection of filter values
+
+Multi select let's users choose more than one value for the same field. The
```
Comment (tomheller) [inline libs/barista-components/filter-field/README.md:200]:
```suggestion
Multi select lets users choose more than one value for the same field. The
```

File: libs/barista-components/filter-field/src/filter-field-module.ts [hunk 4]
```diff
@@ -16,34 +16,36 @@
...
 import { DtLoadingDistractorModule } from '@dynatrace/barista-components/loading-distractor';
 import { DtOverlayModule } from '@dynatrace/barista-components/overlay';
-
 import { DtFilterField } from './filter-field';
+import { DtFilterFieldMultiSelect } from './filter-field-multi-select/filter-field-multi-select';
+import { DtFilterFieldMultiSelectTrigger } from './filter-field-multi-select/filter-field-multi-select-trigger';
 import { DtFilterFieldRange } from './filter-field-range/filter-field-range';
 import { DtFilterFieldRangeTrigger } from './filter-field-range/filter-field-range-trigger';
 import { DtFilterFieldTag } from './filter-field-tag/filter-field-tag';
-import { DtHighlightModule } from '@dynatrace/barista-components/highlight';
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-module.ts:33]: Please consider resetting the import sort.

File: libs/barista-components/filter-field/src/filter-field-module.ts [hunk 5]
```diff
@@ -16,34 +16,36 @@
...
     DtButtonGroupModule,
+    DtButtonModule,
+    DtCheckboxModule,
+    DtHighlightModule,
+    DtIconModule,
+    DtInputModule,
     DtLoadingDistractorModule,
+    DtOptionModule,
     DtOverlayModule,
-    DtHighlightModule,
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-module.ts:46]: Please reset the order to what it was, these sorts of changesets are really hard to decode.

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+        this._scrollToOption();
+      }
+    }
+  }
+
+  /** Determines whether the panel can be opened. */
+  // private _canOpen(): boolean {
+  //   const element = this._elementRef.nativeElement;
+  //   return !element.readOnly && !element.disabled && !this._multiSelectDisabled;
+  // }
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:452]: Code is duplicated below.
```suggestion
```

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+  /**
+   * Strategy that is used to position the panel.
+   */
+  private _positionStrategy: DtFlexibleConnectedPositionStrategy;
+
+  /** Stream of keyboard events that can close the panel. */
+  private readonly _closeKeyEventStream = new Subject<void>();
+
+  /** Subscription to viewport size changes. */
+  // private _viewportSubscription = EMPTY.subscribe();
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:242]: remove unused code.
```suggestion
```

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 8]
```diff
@@ -0,0 +1,758 @@
...
+    return this._multiSelect;
+  }
+  set multiSelect(value: DtFilterFieldMultiSelect<T>) {
+    this._multiSelect = value;
+    this._detachOverlay();
+  }
+  private _multiSelect: DtFilterFieldMultiSelect<T>;
+
+  /** autocomplete` attribute to be set on the input element. */
+  @Input('autocomplete') autocompleteAttribute = 'off';
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:130]: Which autocomplete is this. I don't see why the multi select trigger should require an autocomplete. Maybe you can explain this a bit. 
Imho this trigger does not need a lot more functionality than the range-trigger.

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 9]
```diff
@@ -0,0 +1,758 @@
...
+   * Reference relative to which to position the autocomplete panel.
+   * Defaults to the autocomplete trigger element.
+   */
+  @Input('dtMultiSelectConnectedTo') connectedTo: DtAutocompleteOrigin;
+
+  /**
+   * Reference relative to which to position the autocomplete panel.
+   * Defaults to the autocomplete trigger element.
+   */
+  // @Input('dtMultiSelectConnectedTo') connectedTo: DtAutocompleteOrigin;
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:155]: Unused
```suggestion
```

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 10]
```diff
@@ -0,0 +1,758 @@
...
+      '(multiSelectDisabled || !panelOpen) ? null : multiSelect?.id',
+    '(focusin)': '_handleFocus()',
+    '(blur)': '_handleBlur()',
+    '(input)': '_handleInput($event)',
+    '(keydown)': '_handleKeydown($event)',
+  },
+  providers: [DT_AUTOCOMPLETE_VALUE_ACCESSOR],
+})
+export class DtFilterFieldMultiSelectTrigger<T>
+  implements ControlValueAccessor, OnDestroy {
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:110]: Why does the Multiselect trigger need to implement the ControlValueAccessor? This trigger should function quite the same as the filterfield-range-trigger, but seems far more complicated. 
I don't quite understand why the trigger needs it's own autocomplete that it can disable, and why it needs a controlValueAccessor? 

Maybe you can explain this to me, because I think this can be done similarly to the filter-field-range-trigger.
Please ignore further comments in this file for now.

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+
+/** Injection token to be used to override the default options for `dt-multi-select`. */
+export const DT_MULTI_SELECT_DEFAULT_OPTIONS = new InjectionToken<
+  DtMultiSelectDefaultOptions
+>('dt-multi-select-default-options', {
+  providedIn: 'root',
+  factory: DT_MULTI_SELECT_DEFAULT_OPTIONS_FACTORY,
+});
+
+/** @docs-private */
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts:72]: Has no effect in our repo. please use @internal and prefix it with an `_`
```suggestion
@internal
```

**Review by tomheller — CHANGES_REQUESTED** [review]
A couple of things I noticed: 

- Multi select modal cannot be navigated to or interacted with the keyboard. If a multiselect option is openend, I do think the modal should get the focus and it should be able to check the checkboxes with the keyboard. I know there is a test for it and I didn't check if this one is already passing, but when I tried it, this did not work at all.

- In the multiselect, I'm not sure what the specification says, but I don't think that you should be able to add the same filter twice. Maybe it is possible to remove the "ketchup" option for the second filter (similar to what the autocomplete does). Or is there a specific use case that needs to allow this?
image

- Option filtering by typing: Imho we should consider including the option group as a filterable element as well. When typing home, do I not want to see all condiments within the `Homemade` option group?

- A small bug I noticed when trying to edit in the e2e test app: 
  - Add a Seasoning: Ketchup filter
  - Click on seasoning again, but without completing the filter, try to edit the first filter by clicking on it.
  - The filter field ends up in weird state, where there are 2 flaps open (one is a bit hidden, but you can see it when blurring the filter field.
  - Afterwards, every consecutive filter you want to add in `Seasoning`, will already have ketchup prefilled. It seems there is a check missing if entering the edit mode is possible. 

- When closing the multiselect with `ESC`, there is no way to open it back up again. In the autocomplete, the flap opens again when typing. This does not happen with the multi-select.

-

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts (hunk 10 above)
Comment (tomheller) [reply libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:110]: After thinking about this a bit more, I do think I understand why you need access to the autocomplete. 

In order to filter down the choices that you present in the select, you need access to the input control. I do see that as problematic though, as we now control the autocomplete (input) from two directions. Once from the filterfield itself and once by the multiselect. 
I would suggest finding a way where the filterfield passes the current input value into multiselect, rather than accessing the whole autocomplete. 
Would that make sense?

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    this._checkApplyDisable();
+  }
+
+  /** @internal Toggle option from template */
+  _toggleOptionFromTemplate(event: DtCheckboxChange<T>): void {
+    if (event.checked) this._currentSelection.push(event.source.value);
+    else
+      this._currentSelection = this._currentSelection.filter(
+        (opt) => opt !== event.source.value,
+      );
```
Comment (ffriedl89) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts:300]:
```suggestion
    if (event.checked) {
      this._currentSelection.push(event.source.value);
    } else {
      this._currentSelection = this._currentSelection.filter(
        (opt) => opt !== event.source.value,
      );
    }
```
Please use curly brackets for if statements - this is a potential pitfall

File: libs/barista-components/filter-field/src/filter-field.spec.ts [hunk 13]
```diff
@@ -43,25 +43,100 @@ import {
   FILTER_FIELD_TEST_DATA_SINGLE_DISTINCT,
   FILTER_FIELD_TEST_DATA_SINGLE_OPTION,
 } from '@dynatrace/testing/fixtures';
-import { TEST_DATA_EDITMODE_ASYNC } from './filter-field-edit-mode.spec';
 import {
-  TEST_DATA_SUGGESTIONS,
-  TEST_DATA_EDITMODE,
```
Comment (ffriedl89) [inline libs/barista-components/filter-field/src/filter-field.spec.ts:49]: Why did you move those fixtures into this file?

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 14]
```diff
@@ -656,6 +677,8 @@ export class DtFilterField<T = any>
 
   /** @internal */
   _handleInputKeyDown(event: KeyboardEvent): void {
+    console.log('filter-field', this._loading);
+
```
Comment (ffriedl89) [inline libs/barista-components/filter-field/src/filter-field.ts:681]:
```suggestion

```

**Review by ffriedl89 — CHANGES_REQUESTED** [review]
Thanks @subarroca for putting the work in. I've added a couple of comments regarding the code. The point for improvement is the trigger for me - this is where I see the some potential problems as well, as pointed out by @tomheller. Overall you added a lot tests and examples as well as documentation. Good job :) I think we will find a good solution for the trigger that helps us maintain the different types of overlays a bit better in the future.

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+   * @internal
+   * Manages active item in option list based on key events.
+   */
+  _keyManager: ActiveDescendantKeyManager<DtOption<T>>;
+
+  /** @internal */
+  _portal: TemplatePortal;
+
+  /** @internal Holds the current values of the input field for the from value */
+  _currentSelection: T[] = [];
```
Comment (thomaspink) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select.ts:176]: Angular CDK can help you with selections. It provides a `SelectionModel` class that you can use. Especially powerful for multi selects as you do.
https://github.com/angular/components/blob/master/src/cdk/collections/selection-model.ts

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 16]
```diff
@@ -0,0 +1,758 @@
...
+
+    this._optionHeight = heightConfig.height;
+    this._maxPanelHeight = heightConfig.maxPanelHeight;
+  }
+
+  ngOnDestroy(): void {
+    this._destroy$.next();
+    this._destroy$.complete();
+    this._closingActionsSubscription.unsubscribe();
+    this._windowBlurSubscription.unsubscribe();
```
Comment (thomaspink) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:311]: You already used `.takeUntil(this._destroy$)` for the `_windowBlurSubscription`. So unsubscribing is not necessary.

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts (hunk 10 above)
Comment (thomaspink) [reply libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:110]: As fabian and tom suggested:
- Implementing ControlValueAccessor is not necessary.
- You share a lot of code with the range. Let's create a shared class that both extend from?

**Review by thomaspink — CHANGES_REQUESTED** [review]

**Comment by subarroca** [comment]
> A couple of things I noticed:
> 
> * Multi select modal cannot be navigated to or interacted with the keyboard. If a multiselect option is openend, I do think the modal should get the focus and it should be able to check the checkboxes with the keyboard. I know there is a test for it and I didn't check if this one is already passing, but when I tried it, this did not work at all.
> * In the multiselect, I'm not sure what the specification says, but I don't think that you should be able to add the same filter twice. Maybe it is possible to remove the "ketchup" option for the second filter (similar to what the autocomplete does). Or is there a specific use case that needs to allow this?
>   image
> * Option filtering by typing: Imho we should consider including the option group as a filterable element as well. When typing home, do I not want to see all condiments within the `Homemade` option group?
> * A small bug I noticed when trying to edit in the e2e test app:
>   
>   * Add a Seasoning: Ketchup filter
>   * Click on seasoning again, but without completing the filter, try to edit the first filter by clicking on it.
>   * The filter field ends up in weird state, where there are 2 flaps open (one is a bit hidden, but you can see it when blurring the filter field.
>   * Afterwards, every consecutive filter you want to add in `Seasoning`, will already have ketchup prefilled. It seems there is a check missing if entering the edit mode is possible.
> * When closing the multiselect with `ESC`, there is no way to open it back up again. In the autocomplete, the flap opens again when typing. This does not happen with the multi-select.

I'll address these issues. Thanks for the extensive testing 👍

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts (hunk 8 above)
Comment (subarroca) [reply libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:130]: My mistake. I removed it

**Comment by subarroca** [comment]
TBH The development has been done in 2 stages, one for the functionality and 4 months later for the keyboard interaction. After the first iteration I did an extensive clean-up but when I faced the second one I could not make the keyboard interaction work so I put back lots of code just to achieve it and I did not do the clean-up again.

Thanks for your thorough review and comments. I'll try to address everything in the following week(s)

File: libs/barista-components/filter-field/src/filter-field.spec.ts (hunk 13 above)
Comment (subarroca) [reply libs/barista-components/filter-field/src/filter-field.spec.ts:49]: My fault. It was done some time ago and then I found that you separated the files and tried to keep up to date. I'll put it back

**Comment by subarroca** [comment]
**This is @tomheller  task list**

- [x] Multiselect should be unique by default
- [x] Include option group as a filterable element. (No highlight possible at the moment)
- [ ] Multi select modal cannot be navigated to or interacted with the keyboard. If a multiselect option is openend, I do think the modal should get the focus and it should be able to check the checkboxes with the keyboard. I know there is a test for it and I didn't check if this one is already passing, but when I tried it, this did not work at all.

- [ ] A small bug I noticed when trying to edit in the e2e test app:

- Add a Seasoning: Ketchup filter
- Click on seasoning again, but without completing the filter, try to edit the first filter by clicking on it.
- The filter field ends up in weird state, where there are 2 flaps open (one is a bit hidden, but you can see it when blurring the filter field.
- Afterwards, every consecutive filter you want to add in Seasoning, will already have ketchup prefilled. It seems there is a check missing if entering the edit mode is possible.

- [ ] When closing the multiselect with ESC, there is no way to open it back up again. In the autocomplete, the flap opens again when typing. This does not happen with the multi-select.

**This is @ffriedl89 and @thomaspink list**
- [ ] Remove ControlValueAccessor implementation.
- [ ] Create a  generic class for range and filter and extend them

**Comment by tomheller** [comment]
@subarroca thank you for updating the PR. As of right now, we are quite busy so please be patient as we want to do a thorough review of this PR and this will take some time.

**Review by tomheller — COMMENTED** [review]
Hi @subarroca!

This looks quite good already. Here are a couple of scenarios where I think the behavior could be improved in a second iteration (I don't think this is quite necessary in v1). 

1. When `Month` is selected in the dev app and the focus stays in the input field. `Spacebar` has 2 meanings. First it adds a space to the filtered value, second it would select the checkbox. We might need to do some more focus management here.

2. When selecting `Month` with `ENTER`, then typing `Janu`, then hit `SPACE` and `ENTER`. The selected value is `All`, where only january was filtered. We might need to handle the first active value here a bit differently.

Good work :tada:

**Review by tomheller — DISMISSED** [review]

**Comment by tomheller** [comment]
I have rebased master into the branch and fixed the examples bazel build. 
sadly buildkite is a bit flaky at the moment. The issue has been reported to the responsible team and should be handled shortly.

File: libs/barista-components/filter-field/README.md [hunk 17]
```diff
@@ -103,6 +104,10 @@ Every filter is an array of:
   user. This includes the range operator (`DtFilterFieldRangeOperator`), the
   unit (`string`) of the selected values and the range itself (which can either
   be one `number` value or a `number` tuple).
+- Multi select filters: array of objects that have been selected via an multi
```
Comment (tomheller) [inline libs/barista-components/filter-field/README.md:107]:
```suggestion
- Multi select filters: array of objects that have been selected via a multi
```

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  constructor(
+    protected _elementRef: ElementRef,
+    protected _overlay: Overlay,
+    protected _changeDetectorRef: ChangeDetectorRef,
+    protected _viewportRuler: ViewportRuler,
+    protected _platform: Platform,
+    protected _overlayContainer: OverlayContainer,
+    _zone: NgZone,
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:83]: Nitpick: 
Not quite necessary to prefix this with a `_` as this is not a class member

**Review by tomheller — APPROVED** [review]

File: libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts [hunk 19]
```diff
@@ -80,7 +85,7 @@ export class DtFilterFieldMultiSelectTrigger<T>
     protected _viewportRuler: ViewportRuler,
     protected _platform: Platform,
     protected _overlayContainer: OverlayContainer,
-    _zone: NgZone,
+    protected _zone: NgZone,
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:88]: Sorry, not quite what I meant in the previous review. 
The zone does not need to be a member, because you are simply passing it to the super constructor. 

it's fine either way though.

Comment (subarroca) [reply libs/barista-components/filter-field/src/filter-field-multi-select/filter-field-multi-select-trigger.ts:88]: Yes, I know you meant to remove the underscore but as all the others had this format I preferred adding the `protected`

**Comment by tomheller** [comment]
This looks good now,  buildkite seems to have passed as well. 
@thomaspink & @ffriedl89 you have still requested changes. Could we get this over the finish line?

File: libs/barista-components/filter-field/src/filter-field.html [hunk 20]
```diff
@@ -95,6 +99,19 @@
...
+    <dt-filter-field-multi-select
+      #multiSelect="dtFilterFieldMultiSelect"
+      (multiSelectSubmitted)="_handleMultiSelectSubmitted($event)"
+      [optionsOrGroups]="
+        _currentDef &&
+        !!_currentDef!.multiSelect &&
+        _currentDef!.multiSelect!.multiOptions
+      "
+      [inputValue]="_inputValue"
+      panelWidth="400px"
```
Comment (ffriedl89) [inline libs/barista-components/filter-field/src/filter-field.html:112]: this is gonna be a problem soon, but i am fine with this being in there for v1

**Review by ffriedl89 — APPROVED** [review]
LGTM! Thanks @subarroca for the big big contribution. Ready to merge from my point of view.

File: libs/barista-components/filter-field/src/filter-field.html (hunk 20 above)
Comment (tomheller) [reply libs/barista-components/filter-field/src/filter-field.html:112]: Autocomplete panel is also somewhere clipped. Let's see when it will be a problem.

**Review by tomheller — APPROVED** [review]
